### DISCUSSION TITLE
Improvement ideas for the Ankaios complete state config

### DISCUSSION THREAD

**Discussion by krucod3** [eclipse-ankaios/ankaios#273] [Ideas]
# Background
After some talks with Ankaios users and thoughts on the general usability I would like to collect here ideas on an improved streamlined version of the Ankaios complete state.

# Ideas
Following are some ideas for improvements and not plans to do so:

## desiredState

1. **flatten desired state** - It is planned to remove the startupState from the Complete State, so the question arises if we need the desired state as an object at all, or if we can just go to a flat structure and have the workloads, configs, etc. at the top level.
1. **namespaces for workloads** - If Ankaios manages multiple types of applications, e.g., ADAS, infotainment, etc., it would be nice to have groups of application which are not only differentiated according to tags. This can be achieved by introducing an intermediate level between "workloads:" and the following workload object in the config. The benefit here is that it would be easy to understand and write rules for the operators that manage this applications, i.e., "desiredState.workloads.adas.*" would reference all plications in the adas group. The functionality can also be achieved by adding a prefix to the name of the apps and using the wildcard after the prefix, but it would be more official if there is a defined separation.

1. **"XPath or JSONPath" for the filter masks** @christoph-hamm had an idea to use XPath or JSONPath for the access rights. this will allow a very pinpointed specification of the access including target values

## Workload execution states
1. **filtering execution states* - If we want to allow access to the execution state of certain workload states only according to the workload name or the assigned agent, we would need a way to specify this. The filter mask idea of the initial design will not work here as the execution states are stored in an array. XPath or JSONPath could be used here, but we currently don't have the tags in the execution states. If we copy the tags we would have a lot of redundant information transferred all the time.
1. **Merge the execution states in the workload config** - To make it more convenient to the get the current state, we could merge the execution states and the desired states somehow (as sub-object of workload), but it would be hard to between writable/readable/subscribable information (this is probably a very bad idea)

**Comment by windsource** [comment]
I like the idea of merging the desiredState and the workloadState. This would allow to filter based on all attributes like agent, tags, name etc. 
We can also think of including the workload name in the workload object. Then the workload object would be complete as the name is not only a key but essential part of the workload. On the other hand filtering would be a little bit more complicated as array elements needs to referenced.
